File: 2026-02/src/day_21/boj_3015/Main.java
package day_21.boj_3015;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        Deque<int[]> stack = new ArrayDeque<>();
        long ans = 0;

        int n = Integer.parseInt(br.readLine());
        while (n-- > 0) {
            int x = Integer.parseInt(br.readLine());
            int cnt = 1;

            while (!stack.isEmpty() && stack.peek()[0] < x) {
                ans += stack.pop()[1];
            }

            if (!stack.isEmpty() && stack.peek()[0] == x) {
                int[] top = stack.pop();
                ans += top[1];
                cnt += top[1];
            }

            if (!stack.isEmpty()) ans++;

            stack.push(new int[]{x, cnt});
        }

        System.out.println(ans);
    }
}

File: 2026-02/src/day_21/boj_3015/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    stack<pair<int, int>> st;
    long long ans = 0;

    int n;
    cin >> n;

    while (n--) {
        int x;
        cin >> x;
        int cnt = 1;

        while (!st.empty() && st.top().first < x) {
            ans += st.top().second;
            st.pop();
        }

        if (!st.empty() && st.top().first == x) {
            ans += st.top().second;
            cnt += st.top().second;
            st.pop();
        }

        if (!st.empty()) ans++;

        st.push({x, cnt});
    }

    cout << ans;
}
